### PR TITLE
Add integration tests

### PR DESCRIPTION
**Pull request by zijchen**
This PR adds integration testing based on the Azure Functions in the samples project.

To run the tests, you need to have these installed
- [Azure Functions CLI](https://docs.microsoft.com/azure/azure-functions/functions-run-local)
- [SQL Server Express LocalDB](https://docs.microsoft.com/sql/database-engine/configure-windows/sql-server-express-localdb)

Most of the test infrastructure is in the IntegrationTestBase class. The constructor sets up a test database on LocalDb then starts the Functions Host via Functions CLI. Tests run from there.

TODO: Azure pipelines changes to install the new required tools (unsure if pre-installed), and testing

References
- [Guide for adding Functions integration tests](https://microsoft.github.io/AzureTipsAndTricks/blog/tip196.html)
- [Reference for Functions CLI](https://docs.microsoft.com/en-us/azure/azure-functions/functions-core-tools-reference?tabs=v2#func-start)

### CODE REVIEW

File: test/SqlExtension.IntegrationTests/TestUtils.cs [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+                if (catchException == null || !catchException(ex))
+                {
+                    throw;
+                }
+            }
+            finally
+            {
+                if (cmd != null)
+                {
+                    cmd.Dispose();
```
Comment (Charles-Gagnon) [inline test/SqlExtension.IntegrationTests/TestUtils.cs:120]: Typically it's easier to put the command creation in a using block which will auto-dispose the command after it's done

File: test/SqlExtension.IntegrationTests/SqlExtension.IntegrationTests.csproj [hunk 2]
```diff
@@ -0,0 +1,30 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>netcoreapp3.1</TargetFramework>
+    <IsPackable>false</IsPackable>
```
Comment (Charles-Gagnon) [inline test/SqlExtension.IntegrationTests/SqlExtension.IntegrationTests.csproj:5]: I think it would be better to have all the tests - both unit and integration - in a single project. Splitting them up doesn't seem like something we should need to do and will make code coverage stuff harder.

**Review by Charles-Gagnon — COMMENTED** [review]
There seems to be a lot of other changes in this PR - the samples changes for example. What are those for?

Also - all CS files should have the copyright header at the top

**Comment by Charles-Gagnon** [comment]
That's fine for this time, but in the future yes please try to split stuff up into separate PRs if you can 😄

File: test/SqlExtension.IntegrationTests/SqlExtension.IntegrationTests.csproj (hunk 2 above)
Comment (Charles-Gagnon) [reply test/SqlExtension.IntegrationTests/SqlExtension.IntegrationTests.csproj:5]: Didn't realize you were running the tests on the solution, so the code coverage stuff shouldn't be a problem in that case. But I still think having all the tests in one place makes it easier to work with. 

(although the tests are failing now so you'll need to still do some work there to get them running in the pipeline first. Were you planning on doing that yourself? Maybe we can do that in a follow up PR and just disable them for now in this PR)

Comment (zijchen) [reply test/SqlExtension.IntegrationTests/SqlExtension.IntegrationTests.csproj]: Moved IntegrationTests to inside the Tests project, disabled all but the Ubuntu pipeline for now

File: test/SqlExtension.IntegrationTests/TestUtils.cs (hunk 1 above)
Comment (zijchen) [reply test/SqlExtension.IntegrationTests/TestUtils.cs]: Switched to a using statement

File: builds/azure-pipelines/build-pr.yml [hunk 3]
```diff
@@ -22,10 +22,11 @@ stages:
       matrix:
         linux:
           imageName: 'ubuntu-latest'
-        mac:
-          imageName: 'macos-latest'
-        windows:
-          imageName: 'windows-latest'
+        # Disabling other platform builds temporarily until we can get integration tests working there
```
Comment (Charles-Gagnon) [inline builds/azure-pipelines/build-pr.yml:25]: I would rather we do something else for the platforms that don't support integration tests currently such that the unit tests still run. I'd suggest looking into the filter option to dotnet test - adding a category seems like a decent idea (Categories being Unit/Integration)

File: builds/azure-pipelines/template-steps-build-test.yml [hunk 4]
```diff
@@ -9,6 +9,23 @@ steps:
   inputs:
     version: '3.1.x'
 
+- script: npm install -g azure-functions-core-tools@3 --unsafe-perm true
```
Comment (Charles-Gagnon) [inline builds/azure-pipelines/template-steps-build-test.yml:12]: Why the unsafe perm param?

File: builds/azure-pipelines/template-steps-build-test.yml [hunk 5]
```diff
@@ -9,6 +9,23 @@ steps:
...
+- task: DockerInstaller@0
+  displayName: Docker Installer
+  inputs:
+    dockerVersion: 17.09.0-ce
+    releaseType: stable
+        
+- script: docker pull mcr.microsoft.com/mssql/server:2019-latest
+  displayName: Pull MSSQL Docker Image
+
+- script: 'docker run -e "ACCEPT_EULA=Y" -e "SA_PASSWORD=Placeholder1" -e "MSSQL_PID=Express"
```
Comment (Charles-Gagnon) [inline builds/azure-pipelines/template-steps-build-test.yml:24]: We should generally be avoiding hardcoding passwords. I'd suggest having the pipeline randomly generate one and pass that in instead.

File: samples/SqlExtensionSamples/local.settings.json [hunk 6]
```diff
@@ -0,0 +1,8 @@
+{
```
Comment (Charles-Gagnon) [inline samples/SqlExtensionSamples/local.settings.json:1]: We shouldn't be checking this in - especially for samples that people might be looking at. Something like this should probably be generated by the pipeline itself.

File: test/SqlExtension.Tests/IntegrationTestBase.cs [hunk 7]
```diff
@@ -0,0 +1,252 @@
+﻿using Microsoft.Data.SqlClient;
```
Comment (Charles-Gagnon) [inline test/SqlExtension.Tests/IntegrationTestBase.cs:1]: All new files should have copyright header

File: test/SqlExtension.Tests/IntegrationTestBase.cs [hunk 8]
```diff
@@ -0,0 +1,252 @@
...
+        /// </summary>
+        protected string DatabaseName { get; private set; }
+
+        /// <summary>
+        /// Output redirect for XUnit tests.
+        /// Please use TestOutput.WriteLine() instead of Console or Debug.
+        /// </summary>
+        protected ITestOutputHelper TestOutput { get; private set; }
+
+        protected int Port { get; private set; }
```
Comment (Charles-Gagnon) [inline test/SqlExtension.Tests/IntegrationTestBase.cs:40]: doc comment

File: test/SqlExtension.Tests/IntegrationTestBase.cs [hunk 9]
```diff
@@ -0,0 +1,252 @@
...
+                ExecuteNonQuery(File.ReadAllText(file));
+            }
+
+            // Set SqlConnectionString env var for the Function to use
+            Environment.SetEnvironmentVariable("SqlConnectionString", connectionStringBuilder.ToString());
+        }
+
+        private void StartFunctionHost()
+        {
+            Port = 7071 + TestUtils.ThreadId;
```
Comment (Charles-Gagnon) [inline test/SqlExtension.Tests/IntegrationTestBase.cs:105]: Why do you need to increase this? Shouldn't there only be one function host running at a time?

File: test/SqlExtension.Tests/IntegrationTestBase.cs [hunk 10]
```diff
@@ -0,0 +1,252 @@
...
+        }
+
+        public void Dispose()
+        {
+            Connection.Close();
+
+            try
+            {
+                // Drop the test database
+                using (SqlConnection masterConnection = new SqlConnection("Data Source=(LocalDb)\\MSSQLLocalDB;Initial Catalog=master;Integrated Security=True;Pooling=False;"))
```
Comment (Charles-Gagnon) [inline test/SqlExtension.Tests/IntegrationTestBase.cs:233]: This is still trying to connect to localdb

File: test/SqlExtension.Tests/SqlInputBindingIntegrationTests.cs [hunk 11]
```diff
@@ -0,0 +1,137 @@
+﻿using System;
+using System.Net.Http;
+using System.Text;
+using System.Threading.Tasks;
+using Newtonsoft.Json;
+using SqlExtensionSamples;
+using Xunit;
+using Xunit.Abstractions;
+
+namespace SqlExtension.Tests
```
Comment (Charles-Gagnon) [inline test/SqlExtension.Tests/SqlInputBindingIntegrationTests.cs:10]: It would be convenient to put all the tests into separate folders and under a different namespace based on their type (`SqlExtension.Tests.Integration` and `SqlExtension.Tests.Unit` for example)

File: samples/SqlExtensionSamples/OutputBindingSamples/QueueTriggerProducts.cs [hunk 12]
```diff
@@ -5,22 +5,22 @@
 using Microsoft.Extensions.Logging;
 using static SqlExtensionSamples.ProductUtilities;
 
-namespace SqlExtensionSamples.OutputBindingSamples
+namespace SqlExtensionSamples
```
Comment (Charles-Gagnon) [inline samples/SqlExtensionSamples/OutputBindingSamples/QueueTriggerProducts.cs:8]: Why did you remove the namespace?

File: builds/azure-pipelines/build-pr.yml (hunk 3 above)
Comment (zijchen) [reply builds/azure-pipelines/build-pr.yml]: Added filter to exclude integration tests on Mac and Windows

File: builds/azure-pipelines/template-steps-build-test.yml (hunk 4 above)
Comment (zijchen) [reply builds/azure-pipelines/template-steps-build-test.yml]: Removed (copied from the Functions CLI docs, not sure why it's set there)

File: builds/azure-pipelines/template-steps-build-test.yml (hunk 5 above)
Comment (zijchen) [reply builds/azure-pipelines/template-steps-build-test.yml]: Added a step to generate password using the BuildId

File: test/SqlExtension.Tests/IntegrationTestBase.cs (hunk 9 above)
Comment (zijchen) [reply test/SqlExtension.Tests/IntegrationTestBase.cs]: Removed. Left over code from when I attempted to multithread it

File: samples/SqlExtensionSamples/OutputBindingSamples/QueueTriggerProducts.cs (hunk 12 above)
Comment (zijchen) [reply samples/SqlExtensionSamples/OutputBindingSamples/QueueTriggerProducts.cs:8]: Removed for consistency with the other sample functions, none of the other ones have it

File: samples/SqlExtensionSamples/local.settings.json (hunk 6 above)
Comment (zijchen) [reply samples/SqlExtensionSamples/local.settings.json:1]: Wouldn't we want to include this for people to look at? They will need these settings to run the sample functions (especially the SqlConnectionString which is specific to our samples)

File: samples/SqlExtensionSamples/OutputBindingSamples/AddProduct.cs [hunk 13]
```diff
@@ -9,20 +9,19 @@
...
-            [Sql("Products", ConnectionStringSetting = "SqlConnectionString")] out Product product)
+            [Sql("dbo.Products", ConnectionStringSetting = "SqlConnectionString")] out Product product)
         {
             product = new Product
             {
-                Name = req.Query["name"],
-                ProductID = int.Parse(req.Query["id"]),
-                Cost = int.Parse(req.Query["cost"])
+                Name = req.Query["Name"],
+                ProductID = int.Parse(req.Query["ProductID"]),
```
Comment (chlafreniere) [inline samples/SqlExtensionSamples/OutputBindingSamples/AddProduct.cs:23]: Why are these being changed? Isn't it easier to understand when the params passed in aren't exactly the same as the column names?

Comment (zijchen) [reply samples/SqlExtensionSamples/OutputBindingSamples/AddProduct.cs]: Changed back

File: samples/SqlExtensionSamples/local.settings.json (hunk 6 above)
Comment (Charles-Gagnon) [reply samples/SqlExtensionSamples/local.settings.json:1]: I was basing it off of the guidance from the AF docs about never checking this in. But I guess we have the .gitignore set up properly so maybe it's not the worst idea...seems like other sample guides have it e.g. https://github.com/yokawasa/azure-functions-python-samples/tree/master/v2functions...

I guess as long as it's empty like this it's fine

File: builds/azure-pipelines/template-steps-build-test-ubuntu.yml [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+- task: DockerInstaller@0
+  displayName: Docker Installer
+  inputs:
+    dockerVersion: 17.09.0-ce
+    releaseType: stable
+        
+- script: docker pull mcr.microsoft.com/mssql/server:2019-latest
+  displayName: Pull MSSQL Docker Image
+
+- bash: echo "##vso[task.setvariable variable=serverPassword]Test-$(Build.BuildNumber)"
```
Comment (Charles-Gagnon) [inline builds/azure-pipelines/template-steps-build-test-ubuntu.yml:25]: Since we're doing this already let's at least add in a timestamp or something to randomize it a bit more

File: builds/azure-pipelines/build-pr.yml [hunk 15]
```diff
@@ -14,18 +14,33 @@ stages:
   displayName: 'PR Validation'
 
   jobs:
+  - job: BuildTestOnUbuntu
```
Comment (Charles-Gagnon) [inline builds/azure-pipelines/build-pr.yml:17]: Let's avoid referencing ubuntu specifically for this stuff - just linux is fine

File: samples/SqlExtensionSamples/Product.cs [hunk 16]
```diff
@@ -0,0 +1,11 @@
+﻿namespace SqlExtensionSamples
```
Comment (Charles-Gagnon) [inline samples/SqlExtensionSamples/Product.cs:1]: All new files should have copyright header

**Review by Charles-Gagnon — APPROVED** [review]

**Comment by chlafreniere** [comment]
🔥 🔥 🔥
